File: 664-strange-printer/strange-printer.c
#include <stdio.h>
#include <string.h>

int min(int a, int b) {
    return a < b ? a : b;
}

int strangePrinter(char* s) {
    int n = strlen(s);
    int dp[n][n];
    
    // Initialize the dp array with the worst case
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            dp[i][j] = n;  // Set to maximum possible value initially
        }
        dp[i][i] = 1;  // Single character takes 1 turn
    }
    
    // Fill the dp array for substrings of length greater than 1
    for (int length = 2; length <= n; length++) {
        for (int i = 0; i <= n - length; i++) {
            int j = i + length - 1;
            if (s[i] == s[j]) {
                dp[i][j] = dp[i][j - 1];  // If ends match, it may need fewer turns
            } else {
                for (int k = i; k < j; k++) {
                    dp[i][j] = min(dp[i][j], dp[i][k] + dp[k + 1][j]);
                }
            }
        }
    }
    
    return dp[0][n - 1];
}

File: 664-strange-printer/strange-printer.java
class Solution {
    public int strangePrinter(String s) {
        int n = s.length();
        int[][] dp = new int[n][n];
        
        for (int i = n - 1; i >= 0; i--) {
            dp[i][i] = 1;  // Base case: single character
            for (int j = i + 1; j < n; j++) {
                dp[i][j] = dp[i][j - 1] + 1;  // Worst case, print separately
                
                for (int k = i; k < j; k++) {
                    if (s.charAt(k) == s.charAt(j)) {
                        dp[i][j] = Math.min(dp[i][j], dp[i][k] + dp[k + 1][j - 1]);
                    }
                }
            }
        }
        
        return dp[0][n - 1];
    }
}
